nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -1062,4 +1222,236 @@ mod test {
+            .optional()
+            .unwrap()
+            .is_none()
+    }
+
+    async fn no_region_snapshot_records_exist(datastore: &DataStore) -> bool {
+        use crate::db::model::RegionSnapshot;
+        use crate::db::schema::region_snapshot::dsl;
+
+        dsl::region_snapshot
+            .select(RegionSnapshot::as_select())
+            .first_async::<RegionSnapshot>(
+                datastore.pool_for_tests().await.unwrap(),
+            )
+            .await
+            .optional()
+            .unwrap()
+            .is_none()
+    }
+
+    async fn verify_clean_slate(
+        cptestctx: &ControlPlaneTestContext,
+        test: &DiskTest,
+    ) {
+        // Verifies:
+        // - No disk records exist
+        // - No volume records exist
+        // - No region allocations exist
+        // - No regions are ensured in the sled agent
+        crate::app::sagas::disk_create::test::verify_clean_slate(
+            cptestctx, test,
+        )
+        .await;
+
+        // Verifies:
+        // - No snapshot records exist
+        // - No region snapshot records exist
+        let datastore = cptestctx.server.apictx.nexus.datastore();
+        assert!(no_snapshot_records_exist(datastore).await);
+        assert!(no_region_snapshot_records_exist(datastore).await);
+    }
+
+    #[nexus_test(server = crate::Server)]
+    async fn test_action_failure_can_unwind(
+        cptestctx: &ControlPlaneTestContext,
+    ) {
+        let test = DiskTest::new(cptestctx).await;
+        let log = &cptestctx.logctx.log;
+
+        let client = &cptestctx.external_client;
+        let nexus = &cptestctx.server.apictx.nexus;
+        let mut disk_id = create_org_project_and_disk(&client).await;
+
+        // Build the saga DAG with the provided test parameters
+        let opctx = test_opctx(&cptestctx);
+        let (authz_silo, _authz_org, authz_project, _authz_disk) =
+            LookupPath::new(&opctx, nexus.datastore())
+                .disk_id(disk_id)
+                .lookup_for(authz::Action::Read)
+                .await
+                .expect("Failed to look up created disk");
+
+        let silo_id = authz_silo.id();
+        let project_id = authz_project.id();
+
+        let params = new_test_params(
+            &opctx,
+            silo_id,
+            project_id,
+            disk_id,
+            Name::from_str(DISK_NAME).unwrap().into(),
+        );
+        let mut dag = create_saga_dag::<SagaSnapshotCreate>(params).unwrap();
+
+        // The saga's input parameters include a disk UUID, which makes sense,
+        // since the snapshot is created from a disk.
+        //
+        // Unfortunately, for our idempotency checks, checking for a "clean
+        // slate" gets more expensive when we need to compare region allocations
+        // between the disk and the snapshot. If we can undo the snapshot
+        // provisioning AND delete the disk together, these checks are much
+        // simpler to write.
```

Comment:
This detail matters - basically, it's easy to check "nothing exists in the DB", and harder to check "the state which was provisioned for a disk exists and matches, but nothing else was added".

Unwinding the snapshot *AND* deleting the disk gets us back to the simpler "nothing should exist" state.